text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,107 @@
+    - Setting expiration time to obtain data of a recent period
+    - Setting expiration time to clear inactive data for a specified amount of time
+
+
+### Compatibility
+
+The design adds a VerKV Column Family in RocksDB, which avoids poential compatibility issues with the existing RawKV annd TxnKV data and processing logics, and could be created dynamically when needed. This design allows maximum reuse of the current implementations of TiKV and PD.
+
+### Data model and interface
+
+VerKV is a kv-like interface similar to the RawKV interface. It hides version information from upper-layer services. Its data model is multi-versioned KV between RawKV and TxnKV. 
+
+#### Data model
+
+The data model is like `key:version -> value`. The version uses the current timestamp by default, or can be specified by the user (such as the global `ts` obtained from the PD TSO). And the implementation is subject to further change if we use the user TS feature of RocksDB.
+
+The maximum number of versions in the cluster can be configured via `MaxVerNum`. The default value is 1. Redundant versions are removed during version recycling (compaction or garbage collection).
+
+#### Data storage
+
+`user_key` is encoded according to Memory Comparable Encoding in the TiKV client sdk, and the encoding algorithm is padding in 8-byte units. This operation ensures that ts is appended to the key and the written user_key (key: ts) can keep the same order as the key itself;
+
+The timestamp stored in user_key is its inversion result. The purpose is to arrange the newer data (that is, the data with larger timestamp) in front of the older data (that is, data with smaller timestamp). This feature can be utilized in performance optimizing for the data scan process.
+
+#### Data distribution
+
+To facilitate the acquisition and efficient management of multiple versions of a key, the split point of the region will not use the timestamp of the key, so that multiple versions of the key can be guaranteed to be in the same region and the same CF. Considering the size of the region, MaxVerNum will be restrained.
+
+The process of data balancing, scaling, and migration are the same as that of RawKV. In the process of scanning all versions of a key, if the maximum number of versions is exceeded, the redundant versions will be discarded.
+
+#### RPC Interface
+
+With reference to the proto interface, the following interfaces are defined as a new service other than tikv:
+  - Get/BatchGet -  Gets the values for latest version of the key by default 
+  - GetNum/BatchGetNum - Get the values for a specified number of versions of a key. `version_num` must be less than or equal to `MaxVerNum`. 
+  - GetNumWithSV/BatchGetNumWithSV - This interface is like `GetNum/BatchGetNum`, but add requirement about the key with versions ranging from the specified start version to the current version.
+  - GetWithTTL/BatchGetWithTTL - Gets the unexpired values within the TTL.
```

Comment:
This seems unnecessary given `GetNumWithSV/BatchGetNumWithSV `, is there a difference?

Reply:
`GetWithTTL` is similar with `GetNumWithSV`，but `GetWithTTL` is related with ttl, while `GetNumWithSV` is related with start version.

Reply:
I mean that given that the client should supply the current time, then a TTL limit can be easily computed from the start time and vice versa. (I'm assuming this is using a supplied TTL, not the TTL of the data, which I would expect to be taken into account in any case)